### ISSUE TITLE
Possibly running two instances of dunst

### ISSUE THREAD

**Issue by blueyed** [knopwob/dunst#227]
I've noticed that `dunst` is being run twice for me, apparently because of
using `notify-send` from `~/.xsessionrc`, possibly through `autorandr` and its
postswitch hook.

`/proc/*/environ` appears to be the same for them, but that's likely because it gets launched by `dbus`.

Both processes belong to my user.

The 2nd instance creates theses errors in `journalctl /usr/bin/dbus-daemon`:

    Mai 30 21:35:09 localhost org.freedesktop.Notifications[12712]: BadAccess (attempt to access private resource denied)
    Mai 30 21:35:09 localhost org.freedesktop.Notifications[12712]: BadAccess (attempt to access private resource denied)
    Mai 30 21:35:09 localhost org.freedesktop.Notifications[12712]: Unable to grab key "ctrl+mod1+asciicircum"

I have the following keys configured:

    close = ctrl+space
    close_all = ctrl+shift+space
    history = ctrl+mod1+asciicircum
    context = ctrl+shift+period

I don't think it applies / is relevant here, but I am using some magic to get
notification working in general:

    for X in /tmp/.X11-unix/X*; do
      D="${X##/tmp/.X11-unix/X}"
      user=`who | awk -vD="$D" '$2 ~ ":"D"(.[0-9])?$" {print $1}' | head -1`
      if [ x"$user" != x"" ]; then
        logger "autorandr: Changing display configuration for user '$user'"
        export DISPLAY=":$D"

        HOME=$(getent passwd daniel | cut -d: -f6)
        export XAUTHORITY=$HOME/.Xauthority

        # xrandr -q
        output=$(/bin/su -c "${AUTORANDR}" "$user")
        echo $output

        # Send notification.
        su -c "notify-send --icon=display 'autorandr from $0' '$output'" "$user"

I'll have to investigate how it's triggered, but wanted to provide what I currently know already.

If I understand it correctly, the first instance should register itself on the bus and then be used always?
This would indicate that there are several bus configurations involved?

I am using `gnome-session`, but without `gnome-settings-daemon` etc, and use awesomeWM as window manager.

**Comment by blueyed** [comment]
It's enough to just launch `dunst` from `.xsessionrc` to trigger this.

**Comment by blueyed** [comment]
The session dbus gets launched with the `$STARTUP` script through `/etc/X11/Xsession.d/99x11-common_start`, while `.xsessionrc` gets sourced before (`40x11-common_xsessionrc`).

There's probably not much dunst can do about this, is it?

**Comment by blueyed** [comment]
As a workaround I am using the following wrapper now:

    % cat /usr/local/bin/dunst_singleton
    #!/bin/sh

    pid=$(pgrep --uid $USER -x dunst)
    if [ -n "$pid" ]; then
      kill $pid
    fi

    /usr/local/bin/dunst "$@" &

    if [ -n "$pid" ]; then
      notify-send "Killed old dunst: $pid"
    fi

It is use in the DBUS service file:

    diff --git i/org.knopwob.dunst.service.in w/org.knopwob.dunst.service.in
    index 83c2775..f9eb523 100644
    --- i/org.knopwob.dunst.service.in
    +++ w/org.knopwob.dunst.service.in
    @@ -1,3 +1,3 @@
     [D-BUS Service]
     Name=org.freedesktop.Notifications
    -Exec=##PREFIX##/bin/dunst
    +Exec=##PREFIX##/bin/dunst_singleton

**Comment by hughdavenport** [comment]
Hi,

Is there a second listing of dunst in `ps -C dunst`? If not, not sure dunst can do anything different than giving an error. Prehaps a bug needs to be filed with X11-common?

Cheers,

Hugh

**Comment by blueyed** [comment]
Would `ps -C dunst` be different from `pgrep --uid $USER -x dunst`?

**Comment by hughdavenport** [comment]
`ps -o pid -C dunst` would give same output as `pgrep --uid $USER -x dunst`.  Just wondering whether you have two running at same time, or the second one ran, errored, then quit. If you only see one number in that pgrep listing then the other exited after erroring. If it is, then I would say this isn't a bug with dunst, but with X11-common, as dunst is exiting after an error because it can't run twice.

Cheers,

Hugh

**Comment by blueyed** [comment]
So, yes, I am seeing output / an entry from `pgrep --uid $USER -x dunst` (which gets killed then automatically).
From my understanding, the wrapper script / dunst should not be called if it's running already, and providing the service, correct?

See my previous comment (https://github.com/knopwob/dunst/issues/227#issuecomment-107085149) - it's probably running in different dbus sessions.  I've also noticed that my wrapper kicks in after resuming after hibernation, at least sometimes.

**Comment by hughdavenport** [comment]
Hi,

If it is running in different dbus sessions, I would say this is a bug with X11-common, not dunst. Unless there are two actively running dunst processes running even after one of them throws an error.

Looking at that comment, would it fix your problem if you remove it from .xsessionrc?

Cheers,

Hugh

**Comment by blueyed** [comment]
> Unless there are two actively running dunst processes running even after one of them throws an error.

There are two running dunst instances.  And the first one has bound the keyboard shortcut, but does not appear to respond to it.  The second one does not throw an error, apart from that it cannot bind the shortcut.

The problem seems to be that using `notify-send` in `.xsessionrc` loads it already, but that's before the main dbus session gets started.

**Comment by hughdavenport** [comment]
Are they running at the same time?

One thing I can think of to maybe fix your problem is adding a sleep after the kill in your singleton script to make sure that dunst stops before starting it again.

Hope that helps.

Cheers,

Hugh

**Comment by blueyed** [comment]
> Are they running at the same time?

Yes - if the singleton script would not kill any existing one, i.e. when using dunst as is.

> adding a sleep after the kill in your singleton script 

The singleton script works.  It's just that this is not the default behaviour, and dunst might provide this functionality by itself, i.e. behave like my singleton script and replace any existing instance (from the user) by default?!

**Comment by Eisfreak7** [comment]
I have the same problem. I assume it is caused by using notify-send in systemd units (I have a unit that informs me about priority 0-3 errors in the journal with notify-send). Those are probably started too early. When using notify-send in a terminal after that, I have a second dunst instance which can't be controlled by shortcuts.
```
ps -C dunst
PID TTY          TIME CMD
740 ?        00:00:02 dunst
763 tty1     00:00:02 dunst
```

**Comment by blueyed** [comment]
My workaround (the singleton-wrapper) has the obvious drawback that it will also kill any previous notifications.

Would it be possible for dunst to a) not grab hotkeys with the first instance (detecting that it's not the main D-Bus session somehow), or/and b) to signal the first instance from the second one to move over notifications and/or release grabbed hotkeys etc (and then exit)?

**Comment by Xylemon** [comment]
I am having this two, it's incredibly annoying, especially when notifications cover up each other rather than stack.

**Comment by petr-fischer** [comment]
Hello there - I have also problems with 2 instances of dunst running (under same user). If second instance occur, second instance notification overlap first instance notifications and hotkeys for closing notifications works only for the first instance (most likely).

I have some scripts, running from cron (cronie, arch linux), that emits notifications via notify-send (DISPLAY=:0 is needed for displaying notifications from cron scripts) - is this a problem?

**Comment by Narrat** [comment]
@blueyed: What is the status of this issue after the Arch switch to user dbus-sessions?

And since the stuff is getting called by scripts which seems to miss the dbus env vars.
Maybe the solution from this issue (https://github.com/knopwob/dunst/issues/244) can help?

**Comment by blueyed** [comment]
@Narrat 
Yes, since I've changed to Arch, it's better in this regard, but I had another issue, where it was not started automatically on request through D-Bus (https://github.com/knopwob/dunst/issues/245#issuecomment-158689744).

Currently I start `dunst` explicitly in `~/.xprofile` and source `/etc/X11/xinit/xinitrc.d/?*.sh` in ~/.xinitrc (for the `dbus-update-activation-environment DISPLAY XAUTHORITY`), and had no problems with dunst lately.

**Comment by Narrat** [comment]
What happens if you remove the explicit call in the .xprofile? Or did you try without it after the startup problem was solved?
It shouldn't be necessary.

Else I would suggest to review your open issues. Maybe those vanished with the passed time

**Comment by blueyed** [comment]
> What happens if you remove the explicit call in the .xprofile? Or did you try without it after the startup problem was solved?
> It shouldn't be necessary.

Yes, should work again without (it was my workaround for #245 after all).

Given that it's very similar to #244 - dunst running in different dbus contexts.

I think there's not much dunst can do there really: it's part of the user / system config that should wrap it in the user's session (preferably autostarted).

I am closing the issue for now.
